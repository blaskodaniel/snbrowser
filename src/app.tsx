import React from 'react'
import { Container, CssBaseline, Grid } from '@material-ui/core'
import { Route, Switch } from 'react-router-dom'
import { useRepository } from '@sensenet/hooks-react'
import snLogo from './assets/sensenet_logo_transparent.png'
import { NavBarComponent } from './components/navbar'
import MainPanel from './components/mainpanel'
import DocviewerComponent from './components/document-viewer'
import EditorPage, { EditorPageProps } from './components/document-editor'

/**
 * The main entry point of your app. You can start h@cking from here ;)
 */
export const App: React.FunctionComponent = () => {
  const repo = useRepository()

  return (
    <>
      <CssBaseline />
      <NavBarComponent />
      <Container
        maxWidth="lg"
        style={{
          width: '100%',
          minHeight: '80vh',
          display: 'flex',
          marginTop: '10px',
          verticalAlign: 'middle',
          alignItems: 'center',
          justifyContent: 'flex-start',
          flexDirection: 'column',
          backgroundPosition: 'center',
          backgroundRepeat: 'no-repeat',
          backgroundImage: `url(${snLogo})`,
          backgroundSize: 'auto',
        }}>
        <Grid container>
          <Grid item xs={12}>
            <Switch>
              <Route path="/" exact component={MainPanel} />
              <Route path="/preview/:documentId" component={DocviewerComponent} />
              <Route
                path="/edit/:documentId"
                component={(props: EditorPageProps) => <EditorPage repository={repo} {...props} />}
              />
            </Switch>
          </Grid>
        </Grid>
      </Container>
    </>
  )
}
