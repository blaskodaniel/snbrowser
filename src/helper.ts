export const fakeClick = (obj: EventTarget) => {
  const ev = document.createEvent('MouseEvents')
  ev.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)
  obj.dispatchEvent(ev)
}

export const downloadFile = (name: string, repositoryUrl: string) => {
  const saveLink = document.createElement('a')
  saveLink.href = `${repositoryUrl}${name}?download`
  fakeClick(saveLink)
}
