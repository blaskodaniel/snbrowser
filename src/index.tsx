import React from 'react'
import ReactDOM from 'react-dom'
import { Reducers, Store } from '@sensenet/redux'
import { Provider } from 'react-redux'
import { combineReducers } from 'redux'
import { Repository } from '@sensenet/client-core'
import { Router } from 'react-router-dom'
import { ReduxDiMiddleware } from 'redux-di-middleware'
import './style.css'
import { commentsStateReducer, sensenetDocumentViewerReducer } from '@sensenet/document-viewer-react'
import { App } from './app'
import { RepositoryProvider } from './context/repository-provider'
import history from './utils/browser-history'
import { snInjector } from './context/injector-context-provider'
import { wopi } from './store/wopi-reducer'

/** Initialize the reducers */
const appReducer = combineReducers({
  sensenet: Reducers.sensenet,
  sensenetDocumentViewer: sensenetDocumentViewerReducer,
  comments: commentsStateReducer,
  wopi,
})
export type rootStateType = ReturnType<typeof appReducer>
const diMiddleware = new ReduxDiMiddleware(snInjector)

/** Set the repository */
const repository = new Repository({
  repositoryUrl: 'https://dev.demo.sensenet.com',
})

/** Set the redux store */
const storeOptions = {
  repository,
  middlewares: [diMiddleware.getMiddleware()],
  rootReducer: appReducer,
} as Store.CreateStoreOptions<rootStateType>
const snstore = Store.createSensenetStore(storeOptions)

/**
 * Initialize React
 */
ReactDOM.render(
  /** The RepositoryProvider will display a login form for non-authenticated users */
  <Provider store={snstore}>
    <Router history={history}>
      <RepositoryProvider
        /** You can insert your additional repository settings here */
        sessionLifetime="expiration">
        <App />
      </RepositoryProvider>
    </Router>
  </Provider>,
  document.getElementById('root'),
)
