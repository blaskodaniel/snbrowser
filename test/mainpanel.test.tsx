import { mount } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { Repository } from '@sensenet/client-core'
import { RepositoryContext } from '@sensenet/hooks-react'
import { Breakpoint } from '@material-ui/core/styles/createBreakpoints'
import { ContentList } from '@sensenet/list-controls-react'
import TableRow from '@material-ui/core/TableRow'
import TableBody from '@material-ui/core/TableBody'
import Typography from '@material-ui/core/Typography'
import { CloudDownload, Edit } from '@material-ui/icons'
import Button from '@material-ui/core/Button'
import TableSortLabel from '@material-ui/core/TableSortLabel'
import { downloadFile } from '../src/helper'
import MainPanel from '../src/components/mainpanel'
import { TestContentCollection, TestContentCollectionForOrders } from './_mocks_/test_contents'

jest.mock('../src/helper', () => ({
  downloadFile: jest.fn(),
}))

describe('The main browser panel instance', () => {
  let wrapper: any
  let repo: any

  const MainPanelProps = {
    width: 'lg' as Breakpoint,
  }

  beforeEach(() => {
    window.fetch = function fetchMethod() {
      return Promise.resolve({ d: TestContentCollection } as any)
    }
    repo = new Repository()
    repo.loadCollection = function fetchMethod() {
      return Promise.resolve({ d: { results: TestContentCollection } } as any)
    }
  })

  it('should renders correctly', async () => {
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })

    expect(wrapper.update()).toMatchSnapshot()
  })

  it('should show content list', async () => {
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })

    const contentlist = wrapper.update().find(ContentList)
    expect(contentlist).toBeDefined()

    const tablerow = contentlist.find(TableBody).find(TableRow)
    expect(tablerow.length).toEqual(TestContentCollection.length)
  })

  it('should forward to preview location and back again', async () => {
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })

    const contentlist = wrapper.update().find(ContentList)
    const tablerow = contentlist.find(TableBody).find(TableRow)

    act(() => {
      ;(tablerow.first().prop('onClick') as any)({ target: { innerHTML: TestContentCollection[0].DisplayName } })
    })

    expect(window.location.pathname).toEqual(`/preview/${TestContentCollection[0].Id}`)

    const backbtn = wrapper.find(Button)
    await act(async () => {
      ;(backbtn.first().prop('onClick') as any)()
    })
    const foldernametitle = wrapper.update().find(Typography)
    expect(foldernametitle.text()).toEqual(`/`)
  })

  it('should open a folder', async () => {
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })

    const contentlist = wrapper.update().find(ContentList)
    const tablerow = contentlist.find(TableBody).find(TableRow)

    await act(async () => {
      ;(tablerow.at(1).prop('onDoubleClick') as any)({
        target: { innerHTML: TestContentCollection[1].DisplayName },
      })
    })

    const foldernametitle = wrapper.update().find(Typography)
    expect(foldernametitle.text()).toEqual(`/${TestContentCollection[1].Name}`)
  })

  it('should download a content', async () => {
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })

    const downloadicon = wrapper.update().find(CloudDownload)
    act(() => {
      ;(downloadicon.prop('onClick') as any)()
    })

    expect(downloadFile).toBeCalled()
  })

  it('should open edit mode', async () => {
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })

    const editbtn = wrapper.update().find(Edit)
    act(() => {
      ;(editbtn.first().prop('onClick') as any)({ content: { Id: TestContentCollection[0].Id } })
    })

    expect(window.location.pathname).toEqual(`/edit/${TestContentCollection[0].Id}`)
  })

  it('should order the list', async () => {
    window.fetch = function fetchMethod() {
      return Promise.resolve({ d: TestContentCollectionForOrders } as any)
    }
    repo = new Repository()
    repo.loadCollection = function fetchMethod() {
      return Promise.resolve({ d: { results: TestContentCollectionForOrders } } as any)
    }
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo}>
          <MainPanel {...MainPanelProps} />
        </RepositoryContext.Provider>,
      )
    })
    const contents = wrapper
      .update()
      .find(ContentList)
      .find('tr')

    expect(contents.at(1).text()).toContain(TestContentCollectionForOrders[0].DisplayName)

    const orderbtn = wrapper.update().find(TableSortLabel)
    act(() => {
      ;(orderbtn.at(0).prop('onClick') as any)()
    })
    const orderedcontents = wrapper
      .update()
      .find(ContentList)
      .find('tr')

    const expectedstring = TestContentCollectionForOrders[0].DisplayName
    expect(orderedcontents.at(1).text()).not.toEqual(expectedstring)
  })
})
