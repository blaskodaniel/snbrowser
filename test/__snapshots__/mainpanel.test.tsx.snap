// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The main browser panel instance should renders correctly 1`] = `<WithWidth(MainPanel) />`;
